File: python-modular/examples/all_distance.py
#!/usr/bin/env python
"""
Explicit examples on how to use the different distances
"""

from numpy import array, floor
from numpy.random import seed, rand
from shogun.Features import *
from shogun.Distance import *
from shogun.PreProc import SortWordString

def get_dna ():
	acgt=array(['A', 'C', 'G','T'])
	len_acgt=len(acgt)
	dtrain=[]
	dtest=[]

	for i in xrange(11):
		str1=[]
		str2=[]
		for j in range(60):
			str1.append(acgt[floor(len_acgt*rand())])
			str2.append(acgt[floor(len_acgt*rand())])
		dtrain.append(''.join(str1))
	dtest.append(''.join(str2))
	
	for i in xrange(6):
		str1=[]
		for j in range(60):
			str1.append(acgt[floor(len_acgt*rand())])
	dtest.append(''.join(str1))

	return {'train': dtrain, 'test': dtest}

###########################################################################
# real features
###########################################################################

def euclidian_distance ():
	print 'EuclidianDistance'

	rows=9
	data=rand(rows, 11)
	feats_train=RealFeatures(data)
	data=rand(rows, 19)
	feats_test=RealFeatures(data)

	distance=EuclidianDistance(feats_train, feats_train)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()

def canberra_metric ():
	print 'CanberaMetric'

	rows=9
	data=rand(rows, 10)
	feats_train=RealFeatures(data)
	data=rand(rows, 17)
	feats_test=RealFeatures(data)

	distance=CanberraMetric(feats_train, feats_train)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()

def chebyshew_metric ():
	print 'ChebyshewMetric'

	rows=9
	data=rand(rows, 11)
	feats_train=RealFeatures(data)
	data=rand(rows, 17)
	feats_test=RealFeatures(data)

	distance=ChebyshewMetric(feats_train, feats_train)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()

def geodesic_metric ():
	print 'GeodesicMetric'

	rows=9
	data=rand(rows, 11)
	feats_train=RealFeatures(data)
	data=rand(rows, 21)
	feats_test=RealFeatures(data)

	distance=GeodesicMetric(feats_train, feats_train)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()

def jensen_metric ():
	print 'JensenMetric'

	rows=9
	data=rand(rows, 11)
	feats_train=RealFeatures(data)
	data=rand(rows, 17)
	feats_test=RealFeatures(data)

	distance=JensenMetric(feats_train, feats_train)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()

def manhattan_metric ():
	print 'ManhattanMetric'

	rows=9
	data=rand(rows, 11)
	feats_train=RealFeatures(data)
	data=rand(rows, 17)
	feats_test=RealFeatures(data)

	distance=ManhattanMetric(feats_train, feats_train)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()

def minkowski_metric ():
	print 'MinkowskiMetric'

	rows=9
	data=rand(rows, 11)
	feats_train=RealFeatures(data)
	data=rand(rows, 15)
	feats_test=RealFeatures(data)
	k=3

	distance=MinkowskiMetric(feats_train, feats_train, k)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()

def sparse_euclidian_distance ():
	print 'SparseEuclidianDistance'

	rows=11
	data=rand(rows, 11)
	realfeat=RealFeatures(data)
	feats_train=SparseRealFeatures()
	feats_train.obtain_from_simple(realfeat)
	data=rand(rows, 17)
	realfeat=RealFeatures(data)
	feats_test=SparseRealFeatures()
	feats_test.obtain_from_simple(realfeat)

	distance=SparseEuclidianDistance(feats_train, feats_train)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()


###########################################################################
# complex string features
###########################################################################

def canberra_word_distance ():
	print 'CanberraWordDistance'

	data=get_dna()
	order=3
	gap=0
	reverse=False

	charfeat=StringCharFeatures(DNA)
	charfeat.set_string_features(data['train'])
	feats_train=StringWordFeatures(charfeat.get_alphabet())
	feats_train.obtain_from_char(charfeat, order-1, order, gap, reverse)
	preproc=SortWordString()
	preproc.init(feats_train)
	feats_train.add_preproc(preproc)
	feats_train.apply_preproc()

	charfeat=StringCharFeatures(DNA)
	charfeat.set_string_features(data['test'])
	feats_test=StringWordFeatures(charfeat.get_alphabet())
	feats_test.obtain_from_char(charfeat, order-1, order, gap, reverse)
	feats_test.add_preproc(preproc)
	feats_test.apply_preproc()

	distance=CanberraWordDistance(feats_train, feats_train)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()

def hamming_word_distance ():
	print 'HammingWordDistance'

	data=get_dna()
	order=3
	gap=0
	reverse=False

	charfeat=StringCharFeatures(DNA)
	charfeat.set_string_features(data['train'])
	feats_train=StringWordFeatures(charfeat.get_alphabet())
	feats_train.obtain_from_char(charfeat, order-1, order, gap, reverse)
	preproc=SortWordString()
	preproc.init(feats_train)
	feats_train.add_preproc(preproc)
	feats_train.apply_preproc()

	charfeat=StringCharFeatures(DNA)
	charfeat.set_string_features(data['test'])
	feats_test=StringWordFeatures(charfeat.get_alphabet())
	feats_test.obtain_from_char(charfeat, order-1, order, gap, reverse)
	feats_test.add_preproc(preproc)
	feats_test.apply_preproc()

	use_sign=False

	distance=HammingWordDistance(feats_train, feats_train, use_sign)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()

def manhattan_word_distance ():
	print 'ManhattanWordDistance'

	data=get_dna()
	order=3
	gap=0
	reverse=False

	charfeat=StringCharFeatures(DNA)
	charfeat.set_string_features(data['train'])
	feats_train=StringWordFeatures(charfeat.get_alphabet())
	feats_train.obtain_from_char(charfeat, order-1, order, gap, reverse)
	preproc=SortWordString()
	preproc.init(feats_train)
	feats_train.add_preproc(preproc)
	feats_train.apply_preproc()

	charfeat=StringCharFeatures(DNA)
	charfeat.set_string_features(data['test'])
	feats_test=StringWordFeatures(charfeat.get_alphabet())
	feats_test.obtain_from_char(charfeat, order-1, order, gap, reverse)
	feats_test.add_preproc(preproc)
	feats_test.apply_preproc()

	distance=ManhattanWordDistance(feats_train, feats_train)

	dm_train=distance.get_distance_matrix()
	distance.init(feats_train, feats_test)
	dm_test=distance.get_distance_matrix()

###########################################################################
# call functions
###########################################################################

if __name__=='__main__':
	seed(42)

	euclidian_distance()
	canberra_metric()
	chebyshew_metric()
	geodesic_metric()
	jensen_metric()
	manhattan_metric()
	minkowski_metric()

	sparse_euclidian_distance()

	canberra_word_distance()
	hamming_word_distance()
	manhattan_word_distance()

File: matlab_and_octave/examples/all_distance.m
% Explicit examples on how to use the different distances;

num=24;
len=42;

%
% real features
%

dist=2.4;
traindata_real=[randn(2,num)-dist, randn(2,num)+dist, randn(2,num)+dist*[ones(1,num); zeros(1,num)], randn(2,num)+dist*[zeros(1,num); ones(1,num)]];
testdata_real=[randn(2,num+7)-dist, randn(2,num+7)+dist, randn(2,num+7)+dist*[ones(1,num+7); zeros(1,num+7)], randn(2,num+7)+dist*[zeros(1,num+7); ones(1,num+7)]];

% Euclidian Distance
disp('EuclidianDistance');
sg('send_command', 'set_distance EUCLIDIAN REAL');

sg('set_features', 'TRAIN', traindata_real);;
sg('send_command', 'init_distance TRAIN');
dm=sg('get_distance_matrix');

sg('set_features', 'TEST', testdata_real);;
sg('send_command', 'init_distance TEST');
dm=sg('get_distance_matrix');


% Canberra Metric
disp('CanberaMetric');
sg('send_command', 'set_distance CANBERRA REAL');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', 'init_distance TRAIN');
dm=sg('get_distance_matrix');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_distance TEST');
dm=sg('get_distance_matrix');


% Chebyshew Metric
disp('ChebyshewMetric');
sg('send_command', 'set_distance CHEBYSHEW REAL');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', 'init_distance TRAIN');
dm=sg('get_distance_matrix');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_distance TEST');
dm=sg('get_distance_matrix');


% Geodesic Metric
disp('GeodesicMetric');
sg('send_command', 'set_distance GEODESIC REAL');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', 'init_distance TRAIN');
dm=sg('get_distance_matrix');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_distance TEST');
dm=sg('get_distance_matrix');


% Jensen Metric
disp('JensenMetric');
sg('send_command', 'set_distance JENSEN REAL');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', 'init_distance TRAIN');
dm=sg('get_distance_matrix');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_distance TEST');
dm=sg('get_distance_matrix');


% Manhattan Metric
disp('ManhattanMetric');
sg('send_command', 'set_distance MANHATTAN REAL');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', 'init_distance TRAIN');
dm=sg('get_distance_matrix');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_distance TEST');
dm=sg('get_distance_matrix');


% Minkowski Metric
disp('MinkowskiMetric');
k=3;
sg('send_command', sprintf('set_distance MINKOWSKI REAL %d', k));

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', 'init_distance TRAIN');
dm=sg('get_distance_matrix');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_distance TEST');
dm=sg('get_distance_matrix');


%
% complex string features;
%

order=3;
gap=0;
reverse='n'; % bit silly to not use boolean, set 'r' to yield true

acgt='ACGT';
trainlab_dna=[ones(1,num/2) -ones(1,num/2)];
traindata_dna=acgt(ceil(4*rand(len,num)));
testdata_dna=acgt(ceil(4*rand(len,num)));

% CanberraWord Distance
disp('CanberraWordDistance');

sg('send_command', 'set_distance CANBERRA WORD');
sg('send_command', 'add_preproc SORTWORDSTRING');

sg('set_features', 'TRAIN', traindata_dna, 'DNA');
sg('send_command', sprintf('convert TRAIN STRING CHAR STRING WORD %d %d %d %c', order, order-1, gap, reverse));
sg('send_command', 'attach_preproc TRAIN');
sg('send_command', 'init_distance TRAIN');
dm=sg('get_distance_matrix');

sg('set_features', 'TEST', testdata_dna, 'DNA');
sg('send_command', sprintf('convert TEST STRING CHAR STRING WORD %d %d %d %c', order, order-1, gap, reverse));
sg('send_command', 'attach_preproc TEST');
sg('send_command', 'init_distance TEST');
dm=sg('get_distance_matrix');


% HammingWord Distance
disp('HammingWordDistance');

sg('send_command', 'set_distance HAMMING WORD');
sg('send_command', 'add_preproc SORTWORDSTRING');

sg('set_features', 'TRAIN', traindata_dna, 'DNA');
sg('send_command', sprintf('convert TRAIN STRING CHAR STRING WORD %d %d %d %c', order, order-1, gap, reverse));
sg('send_command', 'attach_preproc TRAIN');
sg('send_command', 'init_distance TRAIN');
dm=sg('get_distance_matrix');

sg('set_features', 'TEST', testdata_dna, 'DNA');
sg('send_command', sprintf('convert TEST STRING CHAR STRING WORD %d %d %d %c', order, order-1, gap, reverse));
sg('send_command', 'attach_preproc TEST');
sg('send_command', 'init_distance TEST');
dm=sg('get_distance_matrix');


% ManhattanWord Distance
disp('ManhattanWordDistance');

sg('send_command', 'set_distance MANHATTAN WORD');
sg('send_command', 'add_preproc SORTWORDSTRING');

sg('set_features', 'TRAIN', traindata_dna, 'DNA');
sg('send_command', sprintf('convert TRAIN STRING CHAR STRING WORD %d %d %d %c', order, order-1, gap, reverse));
sg('send_command', 'attach_preproc TRAIN');
sg('send_command', 'init_distance TRAIN');
dm=sg('get_distance_matrix');

sg('set_features', 'TEST', testdata_dna, 'DNA');
sg('send_command', sprintf('convert TEST STRING CHAR STRING WORD %d %d %d %c', order, order-1, gap, reverse));
sg('send_command', 'attach_preproc TEST');
sg('send_command', 'init_distance TEST');
dm=sg('get_distance_matrix');

